File: mg01/game/tech/Assets/Script/Org/Event/Subject.cs
/* ==============================================================================
 * 观察者
 * @author jr.zeng
 * 2016/6/8 10:40:49
 * ==============================================================================*/

using System;
using System.Linq;
using System.Collections;
using System.Collections.Generic;

using UnityEngine;

namespace mg.org
{

    public class Subject : ISubject
    {
        static ClassPool2<SubjectEvent> __evtPool = ClassPools.me.CreatePool<SubjectEvent>();
        static ClassPool2<Observer> __obsPool = ClassPools.me.CreatePool<Observer>();
        static ClassPool2<List<Observer>> __obsArrPool = ClassPools.me.CreatePool<List<Observer>>();
        
        int m_objNum = 0;
        int m_invalid = 0;

        Dictionary<string, List<Observer>> m_id2obsArr = new Dictionary<string, List<Observer>>();
        static Subject()
        {
            //初始化对象池细节
            __obsPool.capacity = 100;
            __obsPool.chunk = 20;
            __obsArrPool.capacity = 100;
            __obsArrPool.chunk = 20;
        }

        public Subject()
        {

        }


        //-------∽-★-∽------∽-★-∽--------∽-★-∽对象池∽-★-∽--------∽-★-∽------∽-★-∽--------//
        
        //创建obs列表
        List<Observer> CreateObsArr(string type_)
        {
            List<Observer> arr = GetObsArr(type_);
            if (arr == null)
            {
                arr = __obsArrPool.Pop();
                m_id2obsArr[type_] = arr;
            }
            return arr;
        }

        void RemoveObsArr(string type_)
        {
            List<Observer> obsArr;
            if(m_id2obsArr.TryGetValue(type_, out obsArr) )
            {
                obsArr.Clear();
                m_id2obsArr.Remove(type_);
                __obsArrPool.Push(obsArr);

                Log.Debug("回收obsArr:" + __obsArrPool.RemainCount, this);
            }

        }

        //获取obs列表
        List<Observer> GetObsArr(string type_)
        {
            if (!m_id2obsArr.ContainsKey(type_))
                return null;
            return m_id2obsArr[type_];
        }


        //-------∽-★-∽------∽-★-∽--------∽-★-∽事件监听∽-★-∽--------∽-★-∽------∽-★-∽--------//


        /// <summary>
        /// 添加监听
        /// </summary>
        /// <param name="type_"></param>
        /// <param name="callback_"></param>
        /// <param name="refer_"></param>
        public void Attach(string type_, CALLBACK_1 callback_, object refer_)
        {
            if (callback_ == null)
                return;

            if (HasAttach(type_, callback_))
                //已添加监听
                return;

            List<Observer> arr = CreateObsArr(type_);

            Observer obs = __obsPool.Pop();
            obs.Init(type_, callback_, arr, refer_);

            if(arr.Count > 0 && arr[arr.Count-1] == null)
            {
                arr[arr.Count - 1] = obs;
            }
            else
            {
                arr.Add(obs);
            }

            m_objNum++;
        }
        

        /// <summary>
        /// 移除监听
        /// </summary>
        /// <param name="type_"></param>
        /// <param name="callback_"></param>
        /// <returns></returns>
        public void Detach(string type_, CALLBACK_1 callback_)
        {
            List<Observer> arr = GetObsArr(type_);
            if (arr == null )
                return;
            

            if (arr.Count > 0)
            {
                Observer obs;
                for (int i = arr.Count - 1; i >= 0; --i)
                {
                    obs = arr[i];
                    if (obs != null)
                    {
                        if (obs.func == callback_)
                        {
                            obs.Clear();
                            __obsPool.Push(obs);

                            if (m_invalid > 0)
                            {
                                arr[i] = null;
                            }
                            else
                            {
                                arr.RemoveAt(i);
                            }

                            m_objNum--;
                            //break;    可能会有多个
                        }
                    }
                }
            }


            if(arr.Count == 0)
            {
                RemoveObsArr(type_);
            }
            
        }


        /// <summary>
        /// 按类型移除所有监听
        /// </summary>
        /// <param name="type_"></param>
        public void DetachByType(string type_)
        {
            List<Observer> arr = GetObsArr(type_);
            if (arr == null)
                return;
            

            if (arr.Count > 0)
            {
                Observer obs;
                for (int i = arr.Count - 1; i >= 0; --i)
                {
                    obs = arr[i];
                    if (obs != null)
                    {
                        obs.Clear();
                        __obsPool.Push(obs);

                        arr[i] = null;
                        m_objNum--;
                    }
                }
            }
            
            RemoveObsArr(type_);
        }
        

        public void DetachBy_Type_Refer(string type_, object refer_)
        {
            List<Observer> arr = GetObsArr(type_);
            if (arr == null)
                return;

            Refer.Assert(refer_);

            if (arr.Count > 0)
            {
                Observer obs;
                for (int i = arr.Count - 1; i >= 0; --i)
                {
                    obs = arr[i];
                    if (obs != null)
                    {
                        if (obs.refer == refer_)
                        {
                            obs.Clear();
                            __obsPool.Push(obs);

                            if (m_invalid > 0)
                            {
                                arr[i] = null;
                            }
                            else
                            {
                                arr.RemoveAt(i);
                            }

                            m_objNum--;
                            //break;    可能会有多个
                        }
                    }
                }
            }


            if (arr.Count == 0)
            {
                RemoveObsArr(type_);
            }
        }

        public void DetachByRefer(string type_, object refer_)
        {
            //TODO
        }
        /// <summary>
        /// 是否已添加监听
        /// </summary>
        /// <param name="type_"></param>
        /// <param name="callback_"></param>
        /// <returns></returns>
        public bool HasAttach(string type_)
        {
            List<Observer> arr = GetObsArr(type_);
            if (arr == null)
                return false;
            return arr.Count > 0;
        }

        public bool HasAttach(string type_, CALLBACK_1 callback_)
        {
            List<Observer> arr = GetObsArr(type_);
            if (arr == null)
                return false;

            if (arr.Count == 0)
                return false;

            Observer obs;
            for (int i = 0, len = arr.Count; i < len; ++i)
            {
                obs = arr[i];
                if (obs != null)
                {
                    if (obs.func == callback_)
                        return true;
                }
            }
            return false;
        }



      
        //清空观察者
        public void DetachAll()
        {
            if (m_objNum == 0)
                return;
            m_objNum = 0;

            foreach (var kvp in m_id2obsArr)
            {
                if (kvp.Value.Count > 0)
                {
                    foreach (var obs in kvp.Value)
                    {
                        if (obs != null)    //可能刚detach就clear,此时obs会空
                        {
                            obs.Clear();
                            __obsPool.Push(obs);
                        }
                    }
                    kvp.Value.Clear();
                }

                __obsArrPool.Push(kvp.Value);
            }

            m_id2obsArr.Clear();
            m_invalid = 0;
        }



        //-------∽-★-∽------∽-★-∽--------∽-★-∽事件派发∽-★-∽--------∽-★-∽------∽-★-∽--------//


        //通知
        public bool Notify(string type_, object data_)
        {
            List<Observer> arr = GetObsArr(type_);
            if (arr == null)
                return false;
            return Invoke(type_, arr, data_);
        }

        //通知事件
        public bool NotifyEvent(SubjectEvent evt_)
        {
            evt_.curTarget = this;

            List<Observer> arr = GetObsArr(evt_.type);
            if (arr == null)
                return false;
            bool b = InvokeEvent(evt_.type, arr, evt_);

            evt_.curTarget = null;
            return b;
        }

        //使用事件通知
        public bool NotifyWithEvent(string type_, object data_ = null)
        {
            if (!HasAttach(type_))
                return false;

            //SubjectEvent evt = new SubjectEvent(type_, data_);
            SubjectEvent evt = __evtPool.Pop() as SubjectEvent;
            evt.type = type_;
            evt.data = data_;

            bool b = NotifyEvent(evt);
            evt.Clear();

            __evtPool.Push(evt);

            return b;
        }


        bool Invoke(string type_, List<Observer> objArr_, object data_)
        {
            if (objArr_.Count == 0)
                return false;

            ++m_invalid;
            bool succ = false;
            bool b = false;

            Observer obs;
            int i = 0;
            int len = objArr_.Count; //先取长度, 避免新加入的也接收到事件
            for (; i < len; ++i)
            {
                if (objArr_.Count == 0)
                {
                    //$vec被内存回收
                    b = false;
                    break;
                }

                obs = objArr_[i];
                if (obs != null)
                {
                    if(obs.refer == null) 
                    {
                        obs.func(data_);
                        succ = true;
                    }
                    else
                    {
                        //目标在回调时暂时没有作用
                        obs.func(data_);
                        succ = true;
                    }
                    
                }
                else
                {
                    //已经删除
                    b = true;
                }
            }

            if (m_invalid > 0)
                --m_invalid;

            if (b)
            {
                //需要校正
                Adjust(objArr_);

                if (objArr_.Count == 0)
                {
                    RemoveObsArr(type_);
                }
            }

            return succ;
        }

        bool InvokeEvent(string type_, List<Observer> objArr_, SubjectEvent evt_)
        {
            if (objArr_.Count == 0)
                return false;

            ++m_invalid;

            bool succ = false;
            bool b = false;

            Observer obs;
            int i = 0;
            int len = objArr_.Count;
            for (; i < len; ++i)
            {
                if (objArr_.Count == 0)
                {
                    //$vec被内存回收
                    b = false;
                    break;
                }

                obs = objArr_[i];
                if (obs != null)
                {
                    obs.func(evt_);
                    succ = true;
                }
                else
                {
                    //已经删除
                    b = true;
                }

                if (evt_.isStopped)
                    //停止传递
                    break;
            }


            if (m_invalid > 0)
                --m_invalid;

            if (b)
            {
                //需要校正
                Adjust(objArr_);

                if(objArr_.Count == 0)
                {
                    RemoveObsArr(type_);
                }
            }


            return succ;
        }


        void Adjust(List<Observer> objArr_)
        {
            //倒序删除
            for (int i = objArr_.Count - 1; i >= 0; i--)
            {
                if (objArr_[i] == null)
                {
                    objArr_.RemoveAt(i);
                }
            }
        }


        class Observer
        {
            public List<Observer> parent;

            //事件类型
            public string type;
            //回调函数
            public CALLBACK_1 func;
            //引用者
            public object refer;

            public Observer()
            {

            }

            /// <param name="type_"></param>
            /// <param name="func_"></param>
            /// <param name="parent_">此obs所在的队列</param>
            /// <param name="refer_"></param>
            public void Init(string type_, CALLBACK_1 func_, List<Observer> parent_, object refer_)
            {
                type = type_;
                func = func_;
                parent = parent_;
                refer = refer_;

                if (refer != null)
                {
                    //监听目标沉默
                    Refer.AttachDeactive(refer_, onDeacive);
                }
                //else
                //{
                // Refer.Assert(refer_);
                //}
            }

            public void Clear()
            {
                if (refer != null)
                {
                    Refer.DetachDeactive(refer, onDeacive);
                    refer = null;
                }

                type = null;
                func = null;
                parent = null;
            }


            void onDeacive(object referId_)
            {
                //if((string)referId_ != Refer.Format(refer))
                //{
                //    Debug.Assert(false, "不正确的Refer");
                //}
                
                int index = parent.IndexOf(this);
                if (index >= 0)
                {
                    parent[index] = null;   //因为是用置null的方法卸载obs, 所以需要保存parent
                }
                else
                {
                    Log.Assert("错误的序号", this);
                }

                Clear();
                __obsPool.Push(this);
            }

        }

    }


}

File: mg01/game/tech/Assets/Resources/LuaScript/src/org/event/Subject.lua
--观察者模式
--@author jr.zeng
--2017年4月14日 下午4:13:56
local modname = "Subject"
--==================global reference=======================

local table = table
local type = type
local pairs = pairs
local ipairs = ipairs
local assert = assert

--===================namespace========================
local ns = "org"
local using = {--[[namespace1,...]]}
local _ENV = namespace(ns, using, modname)
--===================module property========================
local super = nil	--父类
Subject = class("Subject", super)
local Subject = Subject

local TablePool = TablePool
local clear_tbl = clear_tbl

--Subject.e_pool = Instance(ClassPools):createPool(SubjectEvent)   --事件对象池, 不能挂在这里，会被复制

local evtPool = Inst(ClassPools):createPool(SubjectEvent)
local obsPool = Inst(ClassPools):createPool(Observer)


--创建observer
local function createObs(name, listener, target, notifier, refer)
    local obs = obsPool:pop()
	obs:init(name, listener, target, notifier, refer)
    return obs
end

local function clearObs(obs)
	obs:clear()
	obsPool:push(obs)
end


--===================module content========================

function Subject:__ctor()

    self.m_obsDic = {}
    self.m_obsNum = 0        --监听者的数量

end

function Subject:hasType(type)
    return self.m_obsDic[type] ~= nil
end

--//-------∽-★-∽------∽-★-∽--------∽-★-∽数据管理∽-★-∽--------∽-★-∽------∽-★-∽--------//

--优化点：
--1. obsArr使用对象池
--2. obs使用对象池
--3. 支持单一插槽
--4. 支持refer

--注意
--1. notify中detachByType

--//-------∽-★-∽------∽-★-∽--------∽-★-∽数据操作∽-★-∽--------∽-★-∽------∽-★-∽--------//

--添加监听
--@refer 为false时，显式指定不用refer
--@priority TODO
function Subject:attach(name, listener, target, refer)

    if "function" ~= type(listener) then
        assert(false)
    end

	if refer ~= false then
		refer = refer or target
	end

	local obs

    local isExist = false --查找已有
    local obsArr = self.m_obsDic[name]
    if obsArr then
    
        for k,v in pairs(obsArr) do
            if v.fun == listener and target == v.target then
                isExist = true
                break
            end
        end

        if not isExist then 

	        obs = createObs(name, listener, target, self, refer)
            
--            local l = obsArr[#obsArr]
--            if l then
--                if obs.priority >= l.priority then
--                    obsArr[#obsArr+1] = obs
--                else
--                    table.insert(obsArr, 1, obs)
--                end
--            else
                obsArr[#obsArr+1] = obs
--            end
            
            self.m_obsNum = self.m_obsNum + 1
        end
    else

	    obs = createObs(name, listener, target, self, refer)
	    local obsArr = TablePool.pop()
	    obsArr[#obsArr+1] = obs
        self.m_obsDic[name] = obsArr
        self.m_obsNum = self.m_obsNum + 1
    end

	--return obs
end

--移除监听
function Subject:detach(name, listener, target)

    if "function" ~= type(listener) then
        assert(false)
    end

    local obsArr = self.m_obsDic[name]
    if obsArr == nil  then 
        return end

    for k,obs in ipairs(obsArr) do
        if obs.fun == listener and obs.target == target then
            obsArr[k] = nil
	        clearObs(obs)
            self.m_obsNum = self.m_obsNum - 1
            break
        end
    end    
end


--根据类型移除监听
function Subject:detachByType(type)

    local obsArr = self.m_obsDic[type]
    if obsArr == nil then 
        return end
        
    for k,obs in ipairs(obsArr) do
	    clearObs(obs)
        self.m_obsNum = self.m_obsNum - 1
    end
    self.m_obsDic[type] = nil
    TablePool.push(obsArr)
end


--根据目标移除监听
function Subject:detachByType2(type, target)

    local obsArr = self.m_obsDic[type]
    if obsArr == nil then 
        return end
        
    for k, obs in ipairs(obsArr) do
        if obs.target == target then
            obsArr[k] = nil
	        clearObs(obs)
            self.m_obsNum = self.m_obsNum - 1
        end
    end
end

--根据目标移除监听
function Subject:detachByTarget(target)

    if target == nil then
        return end

    for t, obsArr in pairs(self.m_obsDic) do
        for k, obs in ipairs(obsArr) do
            if obs.target == target then
                obsArr[k] = nil
                clearObs(obs)
                self.m_obsNum = self.m_obsNum - 1
            end
        end
    end
end


function Subject:detachByObs(obs_)

	local name = obs_.name
	local obsArr = self.m_obsDic[obs_.name]
	if not obsArr then
		clearObs(obs_)
		assert(false, '错误的观察者：' .. name)
		return
	end

	--nslog.print_t('detachByObs', obsArr, obs_)

	for k, obs in ipairs(obsArr) do
		if obs == obs_ then

			nslog.print_t(string.format('移除观察者 %s, 事件类型：%s, 剩余数量：%d',
				Refer.format(obs.m_refer),
				name,
				self.m_obsNum-1 ))

			obsArr[k] = nil
			clearObs(obs)
			self.m_obsNum = self.m_obsNum - 1
		end
	end

	--nslog.print(self:dump())
end


--移除所有监听
--@clear 清空观察者
function Subject:detachAll(clear)

    if self.m_obsNum ~= 0 then
        self.m_obsNum = 0
        
        for k,obsArr in pairs(self.m_obsDic) do
            for i, obs in ipairs(obsArr) do
	            clearObs(obs)
            end
            TablePool.push(obsArr)
        end
        clear_tbl(self.m_obsDic)
    end

    --    if clear ~= false then
    --        
    --        self:clearNfyTimeOut() --清空延时派发 
    --    end
end


function Subject:dump()

	local str = string.format('%s, obsNum:%d\n', self.referId, self.m_obsNum )
	for k,obsArr in pairs(self.m_obsDic) do
		str = str ..'type: '.. k .. '\n'
		for i, obs in ipairs(obsArr) do
			str = str ..'\t' .. obs:dump() .. '\n'
		end
	end
	return str
end

--//-------∽-★-∽------∽-★-∽--------∽-★-∽派发相关∽-★-∽--------∽-★-∽------∽-★-∽--------//

local tmp_k = nil

--派发
function Subject:notify(name, ...)

    local obsArr = self.m_obsDic[name]
    if obsArr == nil then 
        return end

    tmp_k = nil
    for k, obs in ipairs(obsArr) do
        if obs.target then
	        obs.fun(obs.target, ...)
        else
	        obs.fun(...)
        end
        tmp_k = k
    end

    if tmp_k == nil then
        --已经没有监听
        self.m_obsDic[name] = nil
        TablePool.push(obsArr)
    end
    tmp_k = nil
end


--派发事件
function Subject:notifyEvent(e)

    assert(e and e.type)

    local obsArr = self.m_obsDic[e.type]
    if obsArr == nil then 
        return end

    e.target = e.target or self

    tmp_k = nil
    for k, obs in ipairs(obsArr) do
        if obs.target then
            obs.fun(obs.target, e)
        else
            obs.fun(e)
        end
        tmp_k = k

        if e.stop then
            --不再往下传递
            break
        end
    end

    if tmp_k == nil then
        --已经没有监听
        self.m_obsDic[e.type] = nil
	    TablePool.push(obsArr)
    end
    tmp_k = nil
end

--用事件派发
function Subject:notifyWithEvent(type, data)

    if self.m_obsDic[type] == nil then 
        return end

    local e = evtPool:pop() --事件可回收
    e.type = type
    e.data = data
    
    self:notifyEvent(e)

    e:clear()
    evtPool:push(e)
end

function Subject:notifyEach(type, fun)

    local obsArr = self.m_obsDic[type]
    if obsArr == nil then
        return end

    tmp_k = nil
    for k, obs in ipairs(obsArr) do
        if obs.target  then
            fun(obs.target, obs.fun)
        else
            fun(nil, obs.fun )
        end
        tmp_k = k
    end

    if tmp_k == nil then
        --已经没有监听
        self.m_obsDic[type] = nil
	    TablePool.push(obsArr)
    end
    tmp_k = nil
end


return Subject